pkg/k8s/watchers/endpoints.go

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package watchers
+
+import (
+	"context"
+	"net/netip"
+	"sync/atomic"
+
+	ipcacheTypes "github.com/cilium/cilium/pkg/ipcache/types"
+	"github.com/cilium/cilium/pkg/k8s"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	"github.com/cilium/cilium/pkg/k8s/watchers/resources"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/source"
+)
+
+func (k *K8sWatcher) newEndpointsInit(swgEps *lock.StoppableWaitGroup) {
+	// Use the K8sAPIGroupEndpointV1Core api group to perform the cache sync wait
+	// regardless of what Resource[*Endpoints] is actually using underneath. This is a
+	// temporary workaround until K8sWatcher and "WaitForCacheSync" have been refactored out.
+	apiGroup := resources.K8sAPIGroupEndpointV1Core
+
+	var synced atomic.Bool
+	synced.Store(false)
+
+	k.blockWaitGroupToSyncResources(
+		k.stop,
+		swgEps,
+		func() bool { return synced.Load() },
+		apiGroup,
+	)
+	k.k8sAPIGroups.AddAPI(apiGroup)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	events := k.sharedResources.Endpoints.Events(ctx)
+	go func() {
+		for {
+			select {
+			case <-k.stop:
+				cancel()
+			case event, ok := <-events:
+				if !ok {
+					return
+				}
+				switch event.Kind {
+				case resource.Sync:
+					synced.Store(true)
+				case resource.Upsert:
+					k.K8sEventReceived(apiGroup, resources.MetricEndpointSlice, resources.MetricUpdate, true, false)
+					k.updateEndpoint(event.Object, swgEps)
+					k.K8sEventProcessed(resources.MetricEndpointSlice, resources.MetricUpdate, true)
+				case resource.Delete:
+					k.K8sEventReceived(apiGroup, resources.MetricEndpointSlice, resources.MetricDelete, true, false)
+					k.K8sSvcCache.DeleteEndpoints(event.Object.EndpointSliceID, swgEps)
+					k.K8sEventProcessed(resources.MetricEndpointSlice, resources.MetricDelete, true)
+				}
+				event.Done(nil)
+			}
+		}
+	}()
+}
+
+func (k *K8sWatcher) updateEndpoint(eps *k8s.Endpoints, swgEps *lock.StoppableWaitGroup) {
+	k.K8sSvcCache.UpdateEndpoints(eps, swgEps)
+	if option.Config.BGPAnnounceLBIP {
+		k.bgpSpeakerManager.OnUpdateEndpoints(eps)
+	}
+	k.addKubeAPIServerServiceEndpoints(eps)
+}
+
+func (k *K8sWatcher) addKubeAPIServerServiceEndpoints(eps *k8s.Endpoints) {
+	if eps == nil ||
+		eps.EndpointSliceID.ServiceID.Name != "kubernetes" ||
+		eps.EndpointSliceID.ServiceID.Namespace != "default" {
+		return
+	}
+	resource := ipcacheTypes.NewResourceID(
+		ipcacheTypes.ResourceKindEndpointSlice,
```

Comment:
Is treating Endpoints as `ipcacheTypes.ResourceKindEndpointSlice` safe regardless of underlying type?

Reply:
This doesn't appear to be persisted to anywhere (checked ipcachemap, kvstore and CiliumEndpoint), so appears to be safe. I've now simplified this to reduce it to just `ResourceKindEndpoint`. This would allow us to switch early to our internal endpoint representation and not carry the concrete endpoint resource kind to everywhere. @joestringer is this OK or am I potentially missing something?